wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityManager.java

```diff
@@ -165,14 +191,276 @@ protected long getIdleTimeout(String wsId) {
   @ScheduleDelay(
       initialDelayParameterName = "che.workspace.activity_check_scheduler_delay_s",
       delayParameterName = "che.workspace.activity_check_scheduler_period_s")
-  private void invalidate() {
+  @VisibleForTesting
+  void validate() {
     try {
-      activityDao.findExpired(System.currentTimeMillis()).forEach(this::stopExpired);
+      stopAllExpired();
+      checkActivityRecordValidity();
```

Comment:
So many lines of code you added in `WorkspaceActivityManager`. Please consider moving it to a dedicated class.